baseplate/lock.py

```diff
@@ -0,0 +1,162 @@
+
+    def acquire(self):
+        """Attempt to acquire the lock.
+
+        CacheLocks also support the `context manager protocol`_, for use with
+        Python's ``with`` statement. When the context is entered, the lock
+        calls :py:meth:`acquire` and when the context is exited it
+        automatically calls :py:meth:`release`.
+
+        :raises: :py:exc:`TimeoutExpired` if ``wait_limit`` is exceeded
+            attempting to get the lock.
+
+        .. _context manager protocol:
+            https://docs.python.org/3/reference/datamodel.html#context-managers
+
+        """
+        self.nonce = "%s.%s" % (socket.gethostname(), os.getpid())
+
+        start = time.time()
+        did_add = False
+        while not did_add:
+            did_add = self.wrapped_cache.add_if_unset(
+                self.key, self.nonce, ttl=self.hold_limit)
+
+            if did_add:
+                self.have_lock = True
+                return
+
+            wait_so_far = time.time() - start
+            remaining_wait_budget = self.wait_limit - wait_so_far
+            if remaining_wait_budget <= 0:
+                raise TimeoutExpired
+            else:
+                # sleep before attempting to get the lock again
+                sleep_time = min(
+                    random.uniform(MIN_RETRY_WAIT, MAX_RETRY_WAIT),
+                    remaining_wait_budget,
+                )
+                time.sleep(sleep_time)
+
+    def release(self):
+        """Release the lock."""
+        if self.wrapped_cache.get(self.key) == self.nonce:
+            self.wrapped_cache.delete(self.key)
+        else:
+            # the lock expired, don't delete someone else's lock
+            pass
+
+        self.have_lock = False
+
+    def __enter__(self):
+        self.acquire()
+        return self
+
+    def __exit__(self, type, value, tb):
+        self.release()
+
+
+def make_lock(cache, key, hold_limit=30, wait_limit=30):
+    """Return a CacheLock for the given key and limits.
+
+    :param cache:
+        pymemcache client instance
+        (:py:class:`pymemcache.client.base.PooledClient`) or
+        redis client instance (:py:class:`redis.StrictRedis`).
+        Typically in baseplate this is an instance of
+        :py:class:`baseplate.context.memcache.MonitoredMemcacheConnection` or
+        :py:class:`baseplate.context.redis.MonitoredRedisConnection`.
+    :param str key: The lock's identifier.
+    :param int hold_limit: The maximum lifetime of the lock, in seconds. The
+        lock is deleted on successful exit, so this is mostly concerned with
+        making sure the lock doesn't persist too long if the delete fails. On
+        the other hand if the lifetime is too short the lock may expire (and
+        be aquireable by other processes) before its owner is finished with it.
+    :param int wait_limit: The maximum time (in seconds) to wait and retry if
+        the lock isn't available.
+
+    """
+
+    if isinstance(cache, PooledClient):
```

Comment:
both of these are relying on the current class hierarchy in the context helpers for the monitored equivalents to continue getting automatically mapped appropriately. that's all internal to baseplate which is fine, but it'd be good to have a test covering that so that if those modules change we notice a test failure.